presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/EquivalenceClassProperty.java

```diff
@@ -0,0 +1,293 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.properties;
+
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.facebook.presto.expressions.LogicalRowExpressions.extractConjuncts;
+import static com.facebook.presto.sql.relational.FunctionResolution.isBuiltInEqualFunction;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Represents classes of equivalent variable and constants references that
+ * hold for a final or intermediate result set produced by a PlanNode.
+ * Variable and constant references are made equivalent via predicate application.
+ * <p>
+ * Each equivalence class is represented by a head. The head is carefully chosen
+ * to be a the member with the strongest binding (i.e. a constant).
+ * <p>
+ * Note that the equivalence class property does not store trivial equivalence
+ * classes with only one member. All variable or constant references are
+ * considered to be in their own virtual equivalence class until combined
+ * into stored larger classes by predicate application.
+ */
+public class EquivalenceClassProperty
+{
+    private final Map<RowExpression, RowExpression> equivalenceClassHeads = new HashMap<>();
+    private final Map<RowExpression, List<RowExpression>> equivalenceClasses = new HashMap<>();
+    private boolean updated;
+
+    public EquivalenceClassProperty() {}
+
+    /**
+     * Determines if one equivalence class property is more general than another.
+     * An equivalence class property is more general than another if it includes all equivalences of the other.
+     *
+     * @param otherEquivalenceClassProperty
+     * @return True if this equivalenceClassProperty is more general than otherEquivalenceClassProperty or False otherwise.
+     */
+    public boolean moreGeneral(EquivalenceClassProperty otherEquivalenceClassProperty)
```

Comment:
probably should name this `isMoreGeneral` or `isMoreGeneralThan`?

Reply:
Done.